terrawrap/utils/config.py

```diff
@@ -86,6 +89,152 @@ def parse_wrapper_configs(wrapper_config_files: List[str]) -> WrapperConfig:
+                post_graph_runs.append(dir_path)
+
+    directory_graph = networkx.compose_all(graph_list)
+    return directory_graph, post_graph_runs
+
+
+# pylint: disable=R0912
+def graph_wrapper_dependencies(config_dir: str, config_dict, graph: networkx.DiGraph, visited: List[str]):
+    """
+    Given a directory, recursively finds all other directories it depends on and builds a graph.
+    :param config_dir: The config directory to obtain a dependency graph for
+    :param config_dict: A dictionary containing wrapper config objects for each seen directory
+    :param graph: The graph to add dependency info to. Empty at first, reused in recursion.
+    :param visited: A list of visited nodes. Empty at first, reused in recursion
+    """
+    if config_dir in visited:
+        return
+    visited.append(config_dir)
+
+    if config_dict.get(config_dir):  # add to dictionary so we only read the file once
+        wrapper_config_obj = config_dict[config_dir].get("wrapper_config")
+    else:
+        wrapper_config_obj = create_wrapper_config_obj(config_dir)
+        config_dict[config_dir] = {
+            "wrapper_config": wrapper_config_obj
+        }
+
+    if wrapper_config_obj.config:
+        graph.add_node(config_dir)
+
+    tf_dependencies = wrapper_config_obj.depends_on
+
+    if tf_dependencies is None:
+        print("Cannot list a dependency without tf_wrapper dependency configuration:", config_dir)
+        sys.exit(1)
+
+    for dependency in tf_dependencies:
+        graph.add_node(dependency)
+        if config_dir in graph:
+            graph.add_edge(dependency, config_dir)
+
+    wrappers = find_wrapper_config_files(config_dir)
+    wrappers.reverse()  # we want the closest wrapper file that gives inherited dependencies
+    for wrapper in wrappers:
+        wrapper_dir = os.path.dirname(wrapper)
+        if config_dict.get(wrapper_dir):  # add to dictionary so we only read the file once
+            new_wrapper_config_obj = config_dict[wrapper_dir].get("wrapper_config")
+        else:
+            new_wrapper_config_obj = create_wrapper_config_obj(wrapper_dir)
+            config_dict[wrapper_dir] = {
+                "wrapper_config": new_wrapper_config_obj
+            }
+
+        if wrapper_dir == config_dir:
+            continue
+        if new_wrapper_config_obj.depends_on is not None:
+            inherited_dependencies = new_wrapper_config_obj.depends_on
+            added = False
+            for dependency in inherited_dependencies:
+                if dependency == config_dir:
+                    continue
+                added = True
+                graph.add_node(dependency)
+                if config_dir in graph:
+                    graph.add_edge(dependency, config_dir)
+            if added:
+                break  # we only need the closest, the recursion will handle anything higher
+
+    for predecessor in list(graph.predecessors(config_dir)):
```

Comment:
since we start at the bottom level we recurse upwards with using the dependency information